File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/MyString.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class MyString {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class MyStringCleaner implements Runnable {

        MemorySegment segment;
        MyStringCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.MyString_destroy(this.segment);
        }
    }

    MyString() {}
    MyString(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static MyString new_(String v) {
        
        try (var arena = Arena.ofConfined()) {
            var vData= Arena.ofAuto().allocateFrom(v, StandardCharsets.UTF_8);
            var vLen = vData.byteSize() - 1;  // allocated strings are null terminated
            var vView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(vView, vLen);
            DiplomatStringView.data(vView, vData);
            var nativeVal = somelib_h.MyString_new(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new MyString(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static MyString newUnsafe(String v) {
        
        try (var arena = Arena.ofConfined()) {
            var vData= Arena.ofAuto().allocateFrom(v, StandardCharsets.UTF_8);
            var vLen = vData.byteSize() - 1;  // allocated strings are null terminated
            var vView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(vView, vLen);
            DiplomatStringView.data(vView, vData);
            var nativeVal = somelib_h.MyString_new_unsafe(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new MyString(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static MyString newFromFirst(String [] v) {
        
        try (var arena = Arena.ofConfined()) {
            var vView = SliceUtils.strs8(Arena.ofAuto(), v);
                var vLen = v.length;
            var nativeVal = somelib_h.MyString_new_from_first(vView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new MyString(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static String stringTransform(String foo) {
        
        try (var arena = Arena.ofConfined()) {
            var fooData= Arena.ofAuto().allocateFrom(foo, StandardCharsets.UTF_8);
            var fooLen = fooData.byteSize() - 1;  // allocated strings are null terminated
            var fooView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(fooView, fooLen);
            DiplomatStringView.data(fooView, fooData);
            var writeable = somelib_h.diplomat_buffer_write_create(0);
            somelib_h.MyString_string_transform(fooView, writeable);
            var string = SliceUtils. readUtf8FromWriteable(writeable);
            somelib_h.diplomat_buffer_write_destroy(writeable);
            return string;
        }
    }
    
    
    public void setStr(String newStr) {
        
        try (var arena = Arena.ofConfined()) {
            
            var newStrData= Arena.ofAuto().allocateFrom(newStr, StandardCharsets.UTF_8);
            var newStrLen = newStrData.byteSize() - 1;  // allocated strings are null terminated
            var newStrView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(newStrView, newStrLen);
            DiplomatStringView.data(newStrView, newStrData);
            somelib_h.MyString_set_str(internal, newStrView);
            
        }
    }
    
    public String getStr() {
        
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.MyString_get_str(internal, writeable);
        var string = SliceUtils. readUtf8FromWriteable(writeable);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/MyString.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface MyStringLib: Library {
    fun MyString_destroy(handle: Pointer)
    fun MyString_new(v: Slice): Pointer
    fun MyString_new_unsafe(v: Slice): Pointer
    fun MyString_new_owned(v: Slice): Pointer
    fun MyString_new_from_first(v: Slice): Pointer
    fun MyString_set_str(handle: Pointer, newStr: Slice): Unit
    fun MyString_get_str(handle: Pointer, write: Pointer): Unit
    fun MyString_get_static_str(): Slice
    fun MyString_string_transform(foo: Slice, write: Pointer): Unit
    fun MyString_borrow(handle: Pointer): Slice
}

class MyString internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class MyStringCleaner(val handle: Pointer, val lib: MyStringLib) : Runnable {
        override fun run() {
            lib.MyString_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<MyStringLib> = MyStringLib::class.java
        internal val lib: MyStringLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun new_(v: String): MyString {
            val vSliceMemory = PrimitiveArrayTools.borrowUtf8(v)
            
            val returnVal = lib.MyString_new(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = MyString(handle, selfEdges)
            CLEANER.register(returnOpaque, MyString.MyStringCleaner(handle, MyString.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newUnsafe(v: String): MyString {
            val vSliceMemory = PrimitiveArrayTools.borrowUtf8(v)
            
            val returnVal = lib.MyString_new_unsafe(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = MyString(handle, selfEdges)
            CLEANER.register(returnOpaque, MyString.MyStringCleaner(handle, MyString.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun newOwned(v: String): MyString {
            val vSliceMemory = PrimitiveArrayTools.moveUtf8(v)
            
            val returnVal = lib.MyString_new_owned(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = MyString(handle, selfEdges)
            CLEANER.register(returnOpaque, MyString.MyStringCleaner(handle, MyString.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun newFromFirst(v: Array<String>): MyString {
            val vSliceMemory = PrimitiveArrayTools.borrowUtf8s(v)
            
            val returnVal = lib.MyString_new_from_first(vSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = MyString(handle, selfEdges)
            CLEANER.register(returnOpaque, MyString.MyStringCleaner(handle, MyString.lib));
            vSliceMemory?.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun getStaticStr(): String {
            
            val returnVal = lib.MyString_get_static_str();
                return PrimitiveArrayTools.getUtf8(returnVal)
        }
        @JvmStatic
        
        fun stringTransform(foo: String): String {
            val fooSliceMemory = PrimitiveArrayTools.borrowUtf8(foo)
            val write = DW.lib.diplomat_buffer_write_create(0)
            val returnVal = lib.MyString_string_transform(fooSliceMemory.slice, write);
            
            val returnString = DW.writeToString(write)
            return returnString
        }
    }
    
    fun setStr(newStr: String): Unit {
        val newStrSliceMemory = PrimitiveArrayTools.borrowUtf8(newStr)
        
        val returnVal = lib.MyString_set_str(handle, newStrSliceMemory.slice);
        
    }
    
    fun getStr(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.MyString_get_str(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    fun borrow(): String {
        
        val returnVal = lib.MyString_borrow(handle);
            return PrimitiveArrayTools.getUtf8(returnVal)
    }

}